dulwich/porcelain.py

```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
             r.refs.set_symbolic_ref(b"HEAD", to_set)
 
 
+def reset_file(repo, file_path: str , target: bytes = b'HEAD'):
+    """
+    reset the file to specific commit or branch
+    Args:
+        repo: dulwich Repo object
+        file_path: file to reset, relative to the repository path
+        target: branch or commit or b'HEAD' to reset
+    """
+    if target in branch_list(repo):
+        sha = repo.refs[b'refs/heads/' + target]
+    elif target == b'HEAD':
+        sha = repo.head()
+    else:
+        sha = target
+    tree = parse_tree(repo, treeish=sha)
+
+    def get_entry(repo, tree, file_path: str):
+        if os.path.split(file_path)[0] != '':  # file_path with directory
+            par_dir_tree_id = get_entry(repo, tree, os.path.split(file_path)[0])[1]
+            tree = parse_tree(repo, treeish=par_dir_tree_id)
+            tree_id = tree[os.path.split(file_path)[1].encode()]
+            return tree_id
+        else:
+            try:
+                return tree[file_path.encode()]
+            except KeyError:
+                # untracked file not in tree.
+                raise KeyError("file '%s' not in tree" % (file_path))
+
+    file_entry = get_entry(repo, tree, file_path)
+    
+    full_path = os.path.join(repo.path, file_path)
+    blob = repo.object_store[file_entry[1]]
+    mode = file_entry[0]
+    build_file_from_blob(blob, mode, full_path.encode())
+
+
+def checkout(repo, branch: bytes, force: bool = False):
+    """
+    switch branches or restore working tree files
+    Args:
+        repo: dulwich Repo object
+        branch: branch name to checkout
+    """
+
+    # check repo status
+    if not force:
+        if status(repo) != GitStatus(staged={'add': [], 'delete': [], 'modify': []}, unstaged=[], untracked=[]):
```

Comment:
What about ignored files? They could be versioned in the target.

Reply:
It seem that ignored files will not appear in the GitSatus, and `git checkout` will not modify the ignored files

Reply:
The branch you're changing to can contain files that were ignored (and not versioned) in the original branch you were on.

Reply:
Does your meaning like this?
```
$ git init
$ # add new thing to file `foo`
$ git add foo; git commit -m 'empty'
$ git branch bran
$ # add new thing to file `bar`
$ # add file `bar` to .gitignore
$ git add .gitignore; git commit -m 'empty1'
```
Here's the output of `git status` in branch `master`:
```
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        bar

nothing added to commit but untracked files present (use "git add" to track)
```

And the output of `git status` in branch `bran`:
```
On branch bran
nothing to commit, working tree clean
```

I think both `git checkout` and  `porcelain.checkout` will not do any thing to the file `bar` after the first commit